crates/noirc_frontend/src/hir_def/types.rs

```diff
@@ -93,26 +188,34 @@ impl std::fmt::Display for StructType {
     }
 }
 
-#[derive(Debug, PartialEq, Eq, Clone)]
+#[derive(Debug, PartialEq, Eq, Clone, Hash)]
 pub enum Type {
     FieldElement(IsConst),
     Array(Box<Type>, Box<Type>),       // Array(4, Field) = [Field; 4]
     Integer(IsConst, Signedness, u32), // u32 = Integer(unsigned, 32)
     PolymorphicInteger(IsConst, TypeVariable),
     Bool(IsConst),
     Unit,
-    Struct(Rc<RefCell<StructType>>, Vec<Type>),
+    Struct(Shared<StructType>, Vec<Type>),
     Tuple(Vec<Type>),
     TypeVariable(TypeVariable),
 
+    /// NamedGenerics are the 'T' or 'U' in a user-defined generic function
+    /// like `fn foo<T, U>(...) {}`. Unlike TypeVariables, they cannot be bound over.
+    NamedGeneric(TypeVariable, Rc<String>),
+
     /// A functions with arguments, a return type, and
     /// a set of possible function ids it may refer to.
     /// The function id set is hidden from users and only
     /// used to monomorphise away most higher order functions.
     Function(Vec<Type>, Box<Type>, BTreeSet<FuncId>),
```

Comment:
It isn't, it was originally meant to be used to monomorphise higher order functions by the function they refer to but it went unused since it wouldn't remove them in all cases. I've removed it.